rest/database_init_manager.go

```diff
@@ -0,0 +1,312 @@
+}
+
+// DatabaseInitOptions specifies the options used for database initialization
+type DatabaseInitOptions struct {
+	indexOptions db.InitializeIndexOptions // Options used for index initialization
+}
+
+func NewDatabaseInitWorker(ctx context.Context, dbName string, n1qlStore *base.ClusterOnlyN1QLStore, collections CollectionInitData, indexOptions db.InitializeIndexOptions, callback collectionCallbackFunc) *DatabaseInitWorker {
+	cancelCtx, cancelFunc := context.WithCancel(ctx)
+	return &DatabaseInitWorker{
+		dbName:                     dbName,
+		options:                    DatabaseInitOptions{indexOptions: indexOptions},
+		ctx:                        cancelCtx,
+		cancelFunc:                 cancelFunc,
+		collections:                collections,
+		n1qlStore:                  n1qlStore,
+		collectionCompleteCallback: callback,
+	}
+}
+
+func (w *DatabaseInitWorker) Run() {
+
+	// Ensure cancelFunc resources are released on normal completion
+	defer func() {
+		if w.cancelFunc != nil {
+			w.cancelFunc()
+		}
+	}()
+
+	// TODO: CBG-2838 refactor initialize indexes to reduce number of system:indexes calls
+	var indexErr error
+	for scName, indexSet := range w.collections {
+		// Add the index set to the common indexOptions
+		collectionIndexOptions := w.options.indexOptions
+		collectionIndexOptions.MetadataIndexes = indexSet
+
+		// TODO: CBG-2838 Refactor InitializeIndexes API to move scope, collection to parameters on system:indexes calls
+		// Set the scope and collection name on the cluster n1ql store for use by initializeIndexes
+		w.n1qlStore.SetScopeAndCollection(scName)
+		indexErr = db.InitializeIndexes(w.ctx, w.n1qlStore, collectionIndexOptions)
+		if indexErr != nil {
+			break
+		}
+
+		// Check for context cancellation after each collection is processed - if cancelled, return cancellation error
+		// to all watchers end exit
+		select {
+		case <-w.ctx.Done():
+			indexErr = errors.New("Database initialization cancelled")
+		default:
+		}
+		if indexErr != nil {
+			break
+		}
+
+		if w.collectionCompleteCallback != nil {
+			w.collectionCompleteCallback(w.dbName, scName.CollectionName())
+		}
+	}
+
+	// On completion (success or error), notify watchers
+	w.watcherLock.Lock()
+	w.lastError = indexErr
+	for _, doneChan := range w.watchers {
+		if indexErr != nil {
+			doneChan <- indexErr
+		}
+		close(doneChan)
+	}
+	w.completed = true
+	w.watcherLock.Unlock()
+}
+
+// Adds a watcher for the current worker.  Creates a new notification channel for completion and adds
+// to watcher set.
+func (w *DatabaseInitWorker) addWatcher() (doneChan chan error, err error) {
+	w.watcherLock.Lock()
+	defer w.watcherLock.Unlock()
+	if w.completed {
+		// If the worker has completed while we acquired the lock, return any error and close channel
+		doneChan = make(chan error, 1)
+		if w.lastError != nil {
+			doneChan <- w.lastError
+		}
+		close(doneChan)
+		return doneChan, nil
```

Comment:
Is it correct to return nil in this case, if know it failed?

Reply:
I think returning nil is correct ("addWatcher" was successful, and the worker status is returned on the doneChan as expected).  Looking at the code I think that means that we never return error from this method, and should just remove the error from the function signature.